packages/headless/src/auth/saml-provider.ts

```diff
@@ -0,0 +1,112 @@
+import {fetch} from 'cross-fetch';
+import {getIsomorphicHistory, IsomorphicHistory} from './isomorphic-history';
+import {getIsomorphicLocation, IsomorphicLocation} from './isomorphic-location';
+
+export interface SamlProviderOptions {
+  organizationId: string;
+  provider: string;
+  request?: Fetch;
+  location?: IsomorphicLocation;
+  history?: IsomorphicHistory;
+}
+
+export interface SamlProvider {
+  exchangeHandshakeToken(): Promise<string>;
+  handshakeTokenAvailable: boolean;
+  login(): void;
+}
+
+type Fetch = (
+  input: RequestInfo,
+  init?: RequestInit | undefined
+) => Promise<Response>;
+
+const handshakeTokenParamName = 'handshake_token';
+
+export function buildSamlProvider(config: SamlProviderOptions): SamlProvider {
+  const options = buildOptions(config);
+  const api = 'https://platform.cloud.coveo.com/rest/search';
+
+  return {
+    login() {
+      const {organizationId, provider, location} = options;
+      const redirectUri = encodeURIComponent(location.href);
+      const params = `organizationId=${organizationId}&redirectUri=${redirectUri}`;
+
+      location.href = `${api}/v2/login/${provider}?${params}`;
+    },
+
+    async exchangeHandshakeToken() {
+      const {location, history, request} = options;
+      const handshakeToken = getHandshakeToken(location);
+      removeHandshakeToken(location, history);
+
+      try {
+        const response = await request(`${api}/login/handshake/token`, {
+          method: 'POST',
+          body: JSON.stringify({handshakeToken}),
+        });
+        const data = await response.json();
+        return data.token;
+      } catch (e) {
+        return '';
+      }
+    },
+
+    get handshakeTokenAvailable() {
+      return !!getHandshakeToken(options.location);
+    },
+  };
+}
+
+function buildOptions(
+  config: SamlProviderOptions
+): Required<SamlProviderOptions> {
+  return {
```

Comment:
Perhaps we should throw an error when `window` is `undefined`.

Reply:
Hmm, if we do this, the unit tests would fail because `window` is not available to them. Similarly, it would prevent someone from server-side rendering a page in a nodejs environment.

SAML auth is only used in a browser, so I think it's ok to do nothing when the code is executed on the server.